### PR TITLE
Added light, temperature and accelerometer on python side

### PR DESCRIPTION
**Pull request by vandyliu**
# Description:

Added light, temperature and accelerometer on python side

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

# Limitations:

Please describe limitations of this PR

# Testing:

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] My code has been formatted with `npm run format` and passes the checks in `npm run check`
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/microbit/__model/display.py [hunk 1]
```diff
@@ -226,13 +227,19 @@ def is_on(self):
...
-
         Use the display's LEDs in reverse-bias mode to sense the amount of light
         falling on the display.  Returns an integer between 0 and 255 representing
         the light level, with larger meaning more light.
         """
-        raise NotImplementedError(CONSTANTS.NOT_IMPLEMENTED_ERROR)
+        return self.__light_level
+
+    def __set_light_level(self, level):
+        if level < CONSTANTS.MIN_LIGHT_LEVEL:
```
Comment (xnkevinnguyen) [inline src/microbit/__model/display.py]: Is this how we should handle out of range values? Or should we throw an error

Comment (vandyliu) [reply src/microbit/__model/display.py]: Great question. Not sure to be honest. On the CPX, if I inputted a value less than the min (on the simulator), it defaulted to the min value, which is why I had it here.

Comment (xnkevinnguyen) [reply src/microbit/__model/display.py]: I see, maybe one of the coaches has advice on this. I am not sure either. I feel like it could "hide" some incorrect inputs if that ever happens

File: src/microbit/test/test_shim.py [hunk 2]
```diff
@@ -20,3 +20,8 @@ def test_running_time(self):
         MicrobitModel.running_time = mock.Mock()
         running_time()
         MicrobitModel.running_time.assert_called_once()
+
```
Comment (andreamah) [inline src/microbit/test/test_shim.py]: Perhaps we should use this PR to consider renaming test_shim.py? Since there's no more shim, the name doesn't really make sense.

File: src/microbit/__model/microbit_model.py [hunk 3]
```diff
@@ -1,23 +1,39 @@
...
         time.sleep(n / 1000)
 
     def running_time(self):
         print(f"time. time: {time.time()}")
         return time.time() - self.__start_time
 
+    def temperature(self):
+        return self.__temperature
+
+    def __set_temperature(self, temperature):
```
Comment (andreamah) [inline src/microbit/__model/microbit_model.py:35]: Considering that you restrain the result here and with light level, perhaps it would be helpful to implement some sort of helper function to restrain it? You can probably put it in common.

File: src/microbit/__model/accelerometer.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
+import enum
```
Comment (andreamah) [inline src/microbit/__model/accelerometer.py]: why do you still import enum here?

**Review by andreamah — COMMENTED** [review]
Yay! Thanks for all this work :D

File: src/microbit/__model/accelerometer.py (hunk 4 above)
Comment (vandyliu) [reply src/microbit/__model/accelerometer.py]: awks, it shall disappear

File: src/microbit/test/test_shim.py (hunk 2 above)
Comment (vandyliu) [reply src/microbit/test/test_shim.py]: changed to test_init.py

File: src/microbit/__model/microbit_model.py (hunk 3 above)
Comment (andreamah) [reply src/microbit/__model/microbit_model.py:35]: https://stackoverflow.com/questions/5996881/how-to-limit-a-number-to-be-within-a-specified-range-python

File: src/microbit/__model/display.py (hunk 1 above)
Comment (isadorasophia) [reply src/microbit/__model/display.py]: I think that we want to do make sense of whatever the real thing does. It wouldn't read a value that is below it's minimum level -- so I'd suggest throwing an exception if that's the case.

Comment (vandyliu) [reply src/microbit/__model/display.py]: Okay, I will update it to raise an exception :)

File: src/microbit/__model/microbit_model.py (hunk 3 above)
Comment (vandyliu) [reply src/microbit/__model/microbit_model.py:35]: I would have done that but the implementation has changed so I won't be using this ):
I'm definitely saving this though! Thank you 💃

**Review by andreamah — APPROVED** [review]
Looks good to me :D

File: src/microbit/__model/accelerometer.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    def was_gesture(self, name):
+        """
+        Return True or False to indicate if the named gesture was active since the
+        last call.
+        """
+        self.__add_current_gesture_to_gesture_lists()
+        if name not in CONSTANTS.GESTURES:
+            raise ValueError(CONSTANTS.INVALID_GESTURE_ERR)
+        was_gesture = name in self.__prev_gestures
+        self.__prev_gestures.clear()
```
Comment (isadorasophia) [inline src/microbit/__model/accelerometer.py:70]: Is there a reason we are clearing this? The API doesn't seem to suggest that we lose the previous state

File: src/microbit/__model/accelerometer.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        elif axis == "y":
+            self.__y = accel
+        elif axis == "z":
+            self.__z = accel
+
+    def __set_gesture(self, gesture):
+        if gesture in CONSTANTS.GESTURES:
+            self.__current_gesture = gesture
+        else:
+            self.__current_gesture = ""
```
Comment (isadorasophia) [inline src/microbit/__model/accelerometer.py:107]: Raise an error?

**Review by isadorasophia — APPROVED** [review]
🏄

File: src/microbit/__model/accelerometer.py (hunk 5 above)
Comment (vandyliu) [reply src/microbit/__model/accelerometer.py:70]: prev_gestures is to remember all gestures that have existed since the last was_gesture call.
From the documentation, once the function is called, all the gestures in prev_gestures disappear (if that makes sense).

File: src/microbit/__model/accelerometer.py (hunk 6 above)
Comment (vandyliu) [reply src/microbit/__model/accelerometer.py:107]: will do

File: src/microbit/test/test_microbit_model.py [hunk 7]
```diff
@@ -20,7 +21,21 @@ def test_running_time(self):
...
+    @pytest.mark.parametrize(
+        "temperature", [CONSTANTS.MIN_TEMPERATURE, 0, CONSTANTS.MAX_TEMPERATURE],
+    )
+    def test_temperature(self, temperature):
+        self.__mb._MicrobitModel__set_temperature(temperature)
+        assert temperature == self.__mb.temperature()
+
+    @pytest.mark.parametrize(
+        "invalid_temperature",
+        [CONSTANTS.MIN_TEMPERATURE - 10, CONSTANTS.MAX_TEMPERATURE + 5],
```
Comment (nasadigital) [inline src/microbit/test/test_microbit_model.py]: _&ast;nit&ast;_
I would do `CONSTANTS.MIN_TEMPERATURE - 1` and `CONSTANTS.MAX_TEMPERATURE + 1` to check the corner cases instead.

**Review by nasadigital — APPROVED** [review]
LGTM ⚡️⚡️⚡️

File: src/microbit/test/test_microbit_model.py (hunk 7 above)
Comment (vandyliu) [reply src/microbit/test/test_microbit_model.py]: will change then merge! Thanks David
